src/math_utils/convolution.py

```
@@ -0,0 +1,23 @@
+    # Validate input
+    if not isinstance(image, Image.Image):
+        raise TypeError("Input must be a PIL Image object")
+    if not isinstance(kernel, np.ndarray) or kernel.ndim != 2:
+        raise TypeError("Kernel must be a 2D NumPy array")
+        
+    image_array = np.array(image.convert("L"))  # grayscale
+    kernel_height, kernel_width = kernel.shape
+    # Calculate padding dynamically based on kernel size
+    pad_h = kernel_height // 2
+    pad_w = kernel_width // 2
+    padded_image = np.pad(image_array, ((pad_h, pad_h), (pad_w, pad_w)), mode='constant')
```

Ruff (0.8.2)

Local variable `padded_image` is assigned to but never used

Remove assignment to unused variable `padded_image`

(F841)